src/core/thread/mle_router.cpp

```diff
@@ -2035,6 +2035,7 @@ ThreadError MleRouter::HandleChildUpdateRequest(const Message &aMessage, const I
     }
 
     child->mLastHeard = Timer::GetNow();
+    child->mAddSrcMatchEntryShort = true;
```

Comment:
I think we need to set `mAddSrcMatchEntryShort` by observing how the child is sending MAC Data Request messages.  If we observe a Data Request with a short source address from the child, then set `mAddSrcMatchEntryShort` to true.  If we observe a Data Request with an extended source address from the child, then set `mAddSrcMatchEntryShort` to false.

This also means we need to cover the case where we need to remove an entry from the short address table and add it to the extended address table and vice-versa.